framework-core/src/main/java/org/daisy/pipeline/event/ProgressMessage.java

```diff
@@ -0,0 +1,685 @@
+package org.daisy.pipeline.event;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.function.Consumer;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import com.google.common.base.Function;
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+
+import org.daisy.common.messaging.Message;
+import org.daisy.common.messaging.MessageAccessor.MessageFilter;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.MDC;
+
+public abstract class ProgressMessage extends Message implements MessageFilter, Iterable<ProgressMessage> {
```

Comment:
This class get pretty long (implementing of several interfaces + contains inner classes), which makes it a bit difficult to read IMO.
Maybe it would benefit from being split in several files? More documentation (in comments?) would help too (e.g. for concepts like active block, nesting, progress portion, etc).